ingest-router/src/routing.rs

```diff
@@ -0,0 +1,216 @@
+                let target = UpstreamTarget {
+                    name: name.clone(),
+                    relay_url: cell_config.relay_url,
+                    sentry_url: cell_config.sentry_url,
+                };
+                (name, target)
+            })
+            .collect();
+        Self { cells }
+    }
+
+    /// Get a specific target by cell name
+    pub fn get(&self, cell_name: &str) -> Option<&UpstreamTarget> {
+        self.cells.get(cell_name)
+    }
+
+    /// Get all cells as a HashMap
+    pub fn cells(&self) -> &HashMap<String, UpstreamTarget> {
+        &self.cells
+    }
+}
+
+/// Routing table that maps locales to their upstream targets
+#[derive(Clone, Debug)]
+pub struct RoutingTable {
+    /// Mapping from locale to cell registry
+    locale_to_cells: HashMap<String, CellRegistry>,
+}
+
+impl RoutingTable {
+    /// Build a routing table from locale cell configurations
+    pub fn new(locales: HashMap<String, HashMap<String, CellConfig>>) -> Self {
+        // Build locale -> cell registry mapping
+        let locale_to_cells = locales
+            .into_iter()
+            .map(|(locale, cells)| {
+                let registry = CellRegistry::from_cells(cells);
+                (locale, registry)
+            })
+            .collect();
+
+        Self { locale_to_cells }
+    }
+
+    /// Get the cell registry for a specific locale
+    pub fn get_cell_registry(&self, locale: &str) -> Option<&CellRegistry> {
+        self.locale_to_cells.get(locale)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_routing_table_creation() {
+        let mut locales = HashMap::new();
+        locales.insert(
+            "us".to_string(),
+            HashMap::from([
+                (
+                    "us1".to_string(),
+                    CellConfig {
+                        sentry_url: Url::parse("http://us1-sentry.example.com").unwrap(),
+                        relay_url: Url::parse("http://us1-relay.example.com").unwrap(),
+                    },
+                ),
+                (
+                    "us2".to_string(),
+                    CellConfig {
+                        sentry_url: Url::parse("http://us2-sentry.example.com").unwrap(),
+                        relay_url: Url::parse("http://us2-relay.example.com").unwrap(),
+                    },
+                ),
+            ]),
+        );
+        locales.insert(
+            "de".to_string(),
+            HashMap::from([(
+                "de".to_string(),
+                CellConfig {
+                    sentry_url: Url::parse("http://de-sentry.example.com").unwrap(),
+                    relay_url: Url::parse("http://de-relay.example.com").unwrap(),
+                },
+            )]),
+        );
+
+        let routing_table = RoutingTable::new(locales);
+
+        // Verify US locale has 2 cells
+        let us_registry = routing_table.get_cell_registry("us").unwrap();
```

Comment:
i'm finding it quite tricky to imagine how the caller is going to use this routing table API. Do you already have some expected usage in mind?

I'm wondering whether there's really value in exposing the cell registry structure and both a cells() and get() api on the locality.

It seems that this function may need to be called from each of the custom actions like `RelayProjectConfigs` (?) so there would be value in making it dead simple for users.

One option i can think of is to simply have 1 function on the `RoutingTable`
```
fn get_upstreams(locality: &str) -> Vec<Upstream>
```